qualtran/serialization/sympy.py

```diff
@@ -0,0 +1,234 @@
+        return sympy_pb2.Parameter(const_int=result)
+
+    # Expression cannot be broken down further, but is a constant.
+    if issubclass(expr.__class__, sympy.core.numbers.Number):
+        if isinstance(expr, sympy.core.numbers.Float):
+            return sympy_pb2.Parameter(const_float=float(expr))
+        if isinstance(expr, sympy.core.numbers.Rational):
+            numerator = _get_sympy_operand(expr.numerator)
+            denominator = _get_sympy_operand(expr.denominator)
+            fraction = sympy_pb2.Rational(numerator=numerator, denominator=denominator)
+            return sympy_pb2.Parameter(const_rat=fraction)
+        else:
+            raise NotImplementedError(f"Sympy expression {str(expr)} cannot be serialized.")
+    if type(expr) is int:
+        return sympy_pb2.Parameter(const_int=expr)
+    if type(expr) is float:
+        return sympy_pb2.Parameter(const_float=expr)
+    else:
+        return _get_const_symbolic_operand(expr)
+
+
+def sympy_expr_to_proto(expr: sympy.Expr) -> sympy_pb2.Term:
+    """Serializes a sympy expression."""
+
+    function = _get_sympy_function_type(expr)
+    operands = []
+    if function == sympy_pb2.Function.NONE:
+        parameter = _get_sympy_operand(expr)
+        operands.append(sympy_pb2.Operand(parameter=parameter))
+
+    else:
+        for term in expr.args:
+            inner_term = sympy_expr_to_proto(term)
+
+            operands.append(sympy_pb2.Operand(term=inner_term))
+
+    return sympy_pb2.Term(function=function, operands=operands)
+
+
+def _get_parameter(
+    serialized_input: Union[sympy_pb2.Operand, sympy_pb2.Parameter]
+) -> Union[sympy.core.AtomicExpr, int, float]:
+    """
+    Deserializes a parameter.
+
+    Deserializes a parameter or operand into either its sympy representation or its
+    python primative numeric representation.
+    """
+    if isinstance(serialized_input, sympy_pb2.Operand):
+        serialized_parameter = serialized_input.parameter
+    else:
+        serialized_parameter = serialized_input
+
+    parameter_type = serialized_parameter.WhichOneof("parameter")
+    if parameter_type == "symbol":
+        return sympy.symbols(serialized_parameter.symbol)
+    if parameter_type == "const_int":
+        return serialized_parameter.const_int
+    if parameter_type == "const_rat":
+        fraction = serialized_parameter.const_rat
+        numerator = _get_parameter(fraction.numerator)
+        denominator = _get_parameter(fraction.denominator)
+        return sympy.Rational(numerator, denominator)
+    if parameter_type == "const_float":
+        return serialized_parameter.const_float
+    if parameter_type == "const_symbol":
+        return _get_sympy_const_from_enum(serialized_parameter.const_symbol)
+
+    raise TypeError(f"Type is not supported for {serialized_input}")
+
+
+def sympy_expr_from_proto(
+    term: sympy_pb2.Term,
+) -> Union[sympy.core.AtomicExpr, sympy_pb2.Term, int, float]:
```

Comment:
The return type should be `sympy.Expr` and not `sympy_pb2.Term` ?

Reply:
This method needs to be able to return both.  When a term consists of other, inner terms, we need to be able to return a sympy_pb2.Term so that we can call this method recursively.  The sympy.Expr type won't always work since not every low-level parameter is a sympy_expression.   Parameters such as "I" are not technically of the type sympy.Expr.  However, both expressions and symbolic constants inherit AtomicExpr.